### DISCUSSION TITLE
How to listen for keyboard events

### DISCUSSION THREAD

**Discussion by runoneall** [fyne-io/fyne#5316] [Q&A] [answered]
I have built an app that consists of a TextGrid, and I want it to listen to keyboard events.

The code is this:
```go
terminal := widget.NewTextGrid()
terminal.SetText("Some Text")
container := fyne_container.NewMax(terminal)
window.SetContent(container)
```

I've tried these, but none of them work:
- SetOnTypedKey
- OnKeyDown

This is the console output:
- `terminal.SetOnTypedKey undefined (type *"fyne.io/fyne/v2/widget".TextGrid has no field or method SetOnTypedKey)`
- `terminal.OnKeyDown undefined (type *"fyne.io/fyne/v2/widget".TextGrid has no field or method OnKeyDown)`

I think this may be an installation problem. I installed it using these commands:
- `go get fyne.io/fyne/v2`
- `go mod tidy`

Here is my complete code:
```go
package main

import (
	"fmt"

	"fyne.io/fyne/v2"
	fyne_app "fyne.io/fyne/v2/app"
	fyne_container "fyne.io/fyne/v2/container"
	fyne_widget "fyne.io/fyne/v2/widget"
)

var input_text = ""

func main() {

	// Window
	app := fyne_app.New()
	window := app.NewWindow("Terminal")
	window.Resize(fyne.NewSize(800, 600))

	// Terminal
	terminal := fyne_widget.NewTextGrid()
	terminal.TabWidth = 4
	terminal.SetText("Some Text")

	// Keyboard Listen (problem)
	//>   terminal.SetOnTypedKey = func(key *fyne.KeyEvent) {
	//>	      fmt.Println(key.Name)
	//>   }

	// Full Window
	container := fyne_container.NewMax(terminal)
	window.SetContent(container)

	// Run
	window.ShowAndRun()
}
```

Who can help me solve this?  Thanks!

**Comment by andydotxyz** [comment, marked as answer]
> This is the console output:
> 
> terminal.SetOnTypedKey undefined (type *"fyne.io/fyne/v2/widget".TextGrid has no field or method SetOnTypedKey)
> terminal.OnKeyDown undefined (type *"fyne.io/fyne/v2/widget".TextGrid has no field or method OnKeyDown)

Correct, those APIs don't exist.

To have a widget respond to events you need to implement the handlers by extending. See docs at:
https://docs.fyne.io/extend/extending-widgets

**Reply by runoneall** [reply]
> > This is the console output:
> > terminal.SetOnTypedKey undefined (type *"fyne.io/fyne/v2/widget".TextGrid has no field or method SetOnTypedKey)
> > terminal.OnKeyDown undefined (type *"fyne.io/fyne/v2/widget".TextGrid has no field or method OnKeyDown)
> 
> Correct, those APIs don't exist.
> 
> To have a widget respond to events you need to implement the handlers by extending. See docs at: https://docs.fyne.io/extend/extending-widgets

Thanks! I solved this problem
